docs/source/publishers.rst

```diff
@@ -0,0 +1,254 @@
+Publishers
+==========
+
+Overview
+--------
+
+Publishers are a framework for transforming alerts prior to dispatching to outputs, on a per-rule basis.
+This allows users to customize the look and feel of alerts.
+
+
+How do Publishers work?
+-----------------------
+
+Publishers are blocks of code that are run during alert processing, immediately prior to dispatching
+an alert to an output.
+
+
+
+Implementing new Publishers
+---------------------------
+
+All publishers must be added to the ``publishers`` directory. Publishers have two valid syntaxes:
+
+
+**Function**
+
+Implement a top-level function with that accepts two arguments: An Alert and a dict. Decorate this function
+with the ``@Register`` decorator.
+
+.. code-block:: python
+
+  @Register
+  def my_publisher(alert: Alert, publication: dict) -> dict:
+    # ...
+    return {}
+
+
+**Class**
+
+Implement a class that inherits from the ``AlertPublisher`` and fill in the implementations for ``publish()``.
+Decorate the class with the ``@Register`` decorator.
+
+.. code-block:: python
+
+  @Register
+  class MyPublisherClass(AlertPublisher):
+
+    def publish(alert: Alert, publication: dict) -> dict:
+      # ...
+      return {}
+
+
+**Recommended Implementation**
+
+Publishers should always return dicts containing only simple types (str, int, list, dict).
+
+Publishers are executed in series, each passing its published ``Alert`` to the next publisher. The ``publication``
```

Comment:
I intended for this to fully error out and blow up the application because—supposing a bug in a publisher—I did not want a bunch of alerts to silently get dispatched with incomplete information. Far safer would be to have the publishers wig out and send the Alert into a retry-loop (until a human stepped in an intervened) which would reduce the change of lost data.

I'm also told that broadly `except Exception:` has to be exercised with caution in python code cuz it can interrupt certain super big bads, like interrupt signals or out-of-memory errors.

As a compromise, I implemented a catch on `KeyError` (since I anticipate this will be like 99% of cases) in `CompositePublisher`. It will error log then reraise the error.